stone/target/js_types.py

```diff
@@ -159,7 +161,35 @@ def _generate_struct(self, struct_type, extra_parameters = [], nameOverride = No
         Emits a JSDoc @typedef for a struct.
         """
         self._emit_jsdoc_header(struct_type.doc)
-        self.emit(' * @typedef {Object} %s' % (nameOverride if nameOverride else fmt_type(struct_type)))
+        self.emit(' * @typedef {Object} %s' % (nameOverride if nameOverride else fmt_type_name(struct_type)))
+
+        # Some structs can explicitly list their subtypes. These structs have a .tag field that indicate
+        # which subtype they are.
+        if struct_type.is_member_of_enumerated_subtypes_tree():
+            if struct_type.has_enumerated_subtypes():
+                # This struct is the parent to multiple subtypes. Determine all of the possible
+                # values of the .tag property.
+                tag_values = []
+                for tags, _ in struct_type.get_all_subtypes_with_tags():
+                    for tag in tags:
+                        tag_values.append('"%s"' % tag)
+
+                jsdoc_tag_union = fmt_jsdoc_union(tag_values)
+                self.emit_wrapped_text('@property {%s} .tag - Tag identifying the subtype variant.' % jsdoc_tag_union)
+            else:
```

Comment:
> It's always only a single root struct and one layer of children

OK; some of the helper functions in the code seemed to assume otherwise, so I went with a general design.

> If one of the subtypes is directly referenced, then there will be no .tag attribute since it isn't ambiguous.

Oh, that complicates things. I could do the following:

1. Make `.tag` an optional field on these data types.
2. Make a second data type for `FilesFileMetadataVariant` or something that has `.tag`, while `FilesFileMetadata` does not.
3. Ignore the `.tag` option entirely.

I think 1 is the cleanest solution. I can put a comment, like "Tag identifying this subtype variant. This field is only present when needed to discriminate between multiple possible types."